src/context/provided_types.rs

```diff
@@ -0,0 +1,92 @@
+use std::{marker::PhantomData, mem, sync::OnceLock};
+
+use better_any::TidAble;
+
+use crate::FileRunContext;
+
+pub trait FromFileRunContextInstanceProvider<'a>: Sized {
+    type Parent: FromFileRunContextInstanceProviderFactory<Provider<'a> = Self>;
+
+    fn get<T: FromFileRunContext<'a> + TidAble<'a>>(
+        &self,
+        file_run_context: FileRunContext<'a, '_, Self::Parent>,
+    ) -> Option<&T>;
+}
+
+pub trait FromFileRunContextInstanceProviderFactory: Send + Sync {
+    type Provider<'a>: FromFileRunContextInstanceProvider<'a, Parent = Self>;
+
+    fn create<'a>(&self) -> Self::Provider<'a>;
+}
+
+pub trait FromFileRunContext<'a> {
+    fn from_file_run_context(
+        file_run_context: FileRunContext<'a, '_, impl FromFileRunContextInstanceProviderFactory>,
+    ) -> Self;
+}
+
+mod _sealed {
+    use super::*;
+
+    pub trait Sealed<'a> {}
+
+    impl<'a, T1: FromFileRunContext<'a>> Sealed<'a> for (T1,) {}
+}
+
+pub trait FromFileRunContextProvidedTypes<'a>: _sealed::Sealed<'a> {
```

Comment:
This seemed appropriate to "seal"

This is the trait that the idea is I will implement for `()`, `(T1)`, `(T1, T2)` etc and that the plugin is expected to declare an implementor of as its "provided types" type

Reply:
Ha, yes I love these weird tuple implementations for things. Which arbitrary number of items in the tuple to stop at?